bpo-41825: restructure docs for the os.wait*() family (#22356)
python/cpython: Doc/library/os.rst

   number is zero); the high bit of the low byte is set if a core file was
   produced.

   If there are no children that could be waited for, :exc:`ChildProcessError`

>    is raised.

Reviewer: I'm surprised `ChildProcessError` wasn't already mentioned, but it's definitely good to see it documented now.

   :func:`waitstatus_to_exitcode` can be used to convert the exit status into an
   exit code.

   .. availability:: Unix.

   .. seealso::

      The other ``wait*`` functions documented below can be used to wait for the
      completion of a specific child process and have more options.
      :func:`waitpid` is the only one also available on Windows.

.. function:: waitid(idtype, id, options)

   Wait for the completion of a child process.  *idtype* can be :data:`P_PID`,
   :data:`P_PGID`, :data:`P_ALL`, or :data:`P_PIDFD` on Linux.  The

>    interpretation of *id* depends on it, see their individual description.

Reviewer: From a formatting perspective, I think it looks clearer to separate the brief description from the parameters.

(Also includes a minor grammar fix)

```suggestion
   Wait for the completion of a child process.  
   
   *idtype* can be :data:`P_PID`, :data:`P_PGID`, :data:`P_ALL`, or :data:`P_PIDFD`
   on Linux.  The interpretation of *id* depends on it, see their individual descriptions.
```

   *options* is an OR combination of flags.  At least one of :data:`WEXITED`,
   :data:`WSTOPPED` or :data:`WCONTINUED` is required.  Additional optional

>    flags are :data:`WNOHANG` and :data:`WNOWAIT`.

Reviewer: I think "At least one or more" is slightly better used here than just "At least one", to very explicitly convey that they can all be used in combination with one another. "one or more" was used in the previous documentation, as well in the man page. While both would be technically correct, IMO "one or more" is a bit more clear.

```suggestion
   *options* is an OR combination of flags.  At least one or more of :data:`WEXITED`,
   :data:`WSTOPPED` or :data:`WCONTINUED` is required.  Additional optional
   flags are :data:`WNOHANG` and :data:`WNOWAIT`.
```

   The return value is an object representing the data contained in the
   :c:type:`siginfo_t` structure with the following attributes: :attr:`si_pid`,

>    :attr:`si_uid`, :attr:`si_signo`, :attr:`si_status`, :attr:`si_code`.

Reviewer: It might be worth expanding here to explain each attribute, or better yet: a specific section for them similar to the `P_*` options. Upon building the documentation, it appears that these sphinx members don't currently link anywhere:

![image](https://user-images.githubusercontent.com/44193521/94389292-7693a900-011d-11eb-9947-cab239c07e59.png)

`si_pid`, `si_uuid`, and `si_signo` are reasonably intuitive, but it would be especially nice to have an explanation for `si_status` (`si_code` could probably just link to the `CLD_*` section). Until consulting the man page, I was personally unaware that it could contain the specific signal that resulted in the child process terminating, stopping, or continuing (my guess was that it only could return the exit status).

Author: Indeed, they are not documented further, I can add that.

   ``-1``, status is requested for any process in the process group ``-pid`` (the
   absolute value of *pid*).

   If :data:`WNOHANG` is specified and there are no matching children in the
   requested state, ``(0, 0)`` is returned.

   If there are no matching children that could be waited for,

>    :exc:`ChildProcessError` is raised.

Reviewer: ```suggestion
   If :data:`WNOHANG` is specified and there are no matching children in the
   requested state, ``(0, 0)`` is returned. If there are no matching children that could be waited for,
   :exc:`ChildProcessError` is raised.
```
This was done for the previous section where `ChildProcessError` was mentioned. IMO, it reads a bit better as a single paragraph.

.. function:: wait4(pid, options)

   Similar to :func:`waitpid`, except a 3-element tuple, containing the child's
   process id, exit status indication, and resource usage information is
   returned.  Refer to :func:`resource.getrusage` for details on resource usage
   information.  The arguments to :func:`wait4` are the same as those provided
   to :func:`waitpid`.

   :func:`waitstatus_to_exitcode` can be used to convert the exit status into an
   exitcode.

   .. availability:: Unix.

.. data:: P_PID
          P_PGID
          P_ALL
          P_PIDFD

   These are the possible values for *idtype* in :func:`waitid`. They affect
   how *id* is interpreted:

   * With :data:`P_PID`, wait for the child with the PID given as *id*.
   * With :data:`P_PGID`, wait for any child whose progress group ID is *id*.
   * With :data:`P_ALL`, wait for any child; *id* is ignored.
   * With :data:`P_PIDFD`, wait for the child identified by the file descriptor
     *id* (a process file descriptor created with :func:`pidfd_open`). This is

>      Linux-specific.

Reviewer: 1) Remove the "With" before each, and replace with just the bullet point and hyphen. This makes it more succinct.

2) Slightly reword `P_PID`. IMO, "wait for the child with the PID given as *id*." would read better as "wait for the child whose PID is *id*".

```suggestion
   * :data:`P_PID` - wait for the child whose PID is *id*.
   * :data:`P_PGID` - wait for any child whose progress group ID is *id*.
   * :data:`P_ALL` - wait for any child; *id* is ignored.
   * :data:`P_PIDFD` - wait for the child identified by the file descriptor
     *id* (a process file descriptor created with :func:`pidfd_open`).
```

Author: Nice!

   .. availability:: Unix, ``P_PIDFD`` only on Linux 5.4+.

   .. versionadded:: 3.3
   .. versionadded:: 3.9
      The ``P_PIDFD`` constant.

.. data:: WCONTINUED

   This option for :func:`waitpid`, :func:`wait3`, :func:`wait4`, and
   :func:`waitid` causes child processes to also be reported if they have been
   continued from a job control stop since their status was last reported.

>    .. availability:: some Unix systems.

Reviewer: On Linux, `WCONTINUED` is specifically available on kernel 2.6.10+. However, I'm not certain that it's worth mentioning explicitly since 2.6.x has been EoL for a while now.

   by the delivery of a signal to be reported.

   This option is not available for the other ``wait*`` functions.

   .. availability:: Unix.

   .. versionadded:: 3.3

.. data:: WUNTRACED

   This option for :func:`waitpid`, :func:`wait3`, and :func:`wait4` causes
   child processes to also be reported if they have been stopped but their
   current state has not been reported since they were stopped.

   This option is not available for :func:`waitid`.

   .. availability:: Unix.

.. data:: WNOHANG

   The option for :func:`waitpid`, :func:`wait3`, :func:`wait4`, and
   :func:`waitid` to return immediately if no child process status is available
   immediately.

   .. availability:: Unix.

.. data:: WNOWAIT

   The option for :func:`waitid` to leave the child in a waitable state, so that
   a later wait call can be used again to retrieve the child status information.

   This option is not available for the other ``wait*`` functions.

   .. availability:: Unix.

.. data:: CLD_EXITED
          CLD_KILLED
          CLD_DUMPED
          CLD_TRAPPED
          CLD_STOPPED
          CLD_CONTINUED

   These are the possible values for :attr:`si_code` in the result returned by

>    :func:`waitid`.

Reviewer: Would it be worth including some brief details on each `si_code` value? 

I don't think it would be quite as useful as the `P_*` options, but some are not entirely intuitive. Particularly for `CLD_EXITED`, I would not guess that it's very specifically denoting that the child process called `_exit(2)` without consulting documentation. Others such as `CLD_KILLED`, `CLD_STOPPED`, and `CLD_CONTINUED` are significantly more obvious based on the name, but I think it would still be useful to have a short description with sphinx links to the `signal` module.

Author: > I would not guess that it's very specifically denoting that the child process called _exit(2) without consulting documentation

I don't think that's as significant as you think, since `_exit` is also what the C runtime calls when the `main` function finishes normally. So it really is just "exited normally".

   number is zero); the high bit of the low byte is set if a core file was
   produced.

   If there are no children that could be waited for, :exc:`ChildProcessError`
   is raised.

   :func:`waitstatus_to_exitcode` can be used to convert the exit status into an
   exit code.

   .. availability:: Unix, not Emscripten, not WASI.

   .. seealso::

>       The other ``wait*`` functions documented below can be used to wait for the

Reviewer 2: To use consistent syntax, formatting and semantics as other functions, but not try to resolve the reference, you could do:

```suggestion
      The other :func:`!wait*` functions documented below can be used to wait for the
```

Or, if you wanted to make it more explict that the `*` reprisents a wildcard, you could do

```suggestion
      The other :samp:`wait{*}` functions documented below can be used to wait for the
```

      completion of a specific child process and have more options.
      :func:`waitpid` is the only one also available on Windows.

.. function:: waitid(idtype, id, options, /)

   Wait for the completion of a child process.

   *idtype* can be :data:`P_PID`, :data:`P_PGID`, :data:`P_ALL`, or :data:`P_PIDFD`

>    on Linux.  The interpretation of *id* depends on it, see their individual descriptions.

Reviewer 2: ```suggestion
   on Linux.  The interpretation of *id* depends on it; see their individual descriptions.
```

Fix grammar error (comma splice)

>    on Linux.  The interpretation of *id* depends on it, see their individual descriptions.

Reviewer 2: This states that *idtype* can be one of these options on Linux, but what about other *nix (macOS, BSD) which this function says it supports? The `availability` on these constants below also says `Unix`, which isn't consistent with just "Linux" stated here, so presumably one, the other or both should be clarified?

Author: I think I meant `(a, b, c) or (d on Linux)`.
```suggestion
   *idtype* can be :data:`P_PID`, :data:`P_PGID`, :data:`P_ALL`, or, on Linux, :data:`P_PIDFD`.
   The interpretation of *id* depends on it; see their individual descriptions.
```

>    *options* is an OR combination of flags.  At least one or more of :data:`WEXITED`,

Reviewer 2: "At least one or more" is redundant, since "at least N" and "N or more" are synonymous. So, this should be either

```suggestion
   *options* is an OR combination of flags.  At least one of :data:`WEXITED`,
```

or

```suggestion
   *options* is an OR combination of flags.  One or more of :data:`WEXITED`,
```

   :data:`WSTOPPED` or :data:`WCONTINUED` is required.  Additional optional

>    flags are :data:`WNOHANG` and :data:`WNOWAIT`.

Reviewer 2: ```suggestion
   :data:`WSTOPPED` or :data:`WCONTINUED` is required;
   :data:`WNOHANG` and :data:`WNOWAIT` are additional optional flags.
```

IMO, this reduces excessive choppiness and improves flow.

   The return value is an object representing the data contained in the
   :c:type:`siginfo_t` structure with the following attributes: :attr:`si_pid`
   (process ID), :attr:`si_uid` (real user ID of the child), :attr:`si_signo`
   (always :data:`~signal.SIGCHLD`), :attr:`si_status` (the exit status or
   signal number, depending on :attr:`si_code`), and :attr:`si_code` (see

>    :data:`CLD_EXITED` for possible values).

Reviewer 2: ```suggestion
   The return value is an object representing the data contained in the
   :c:type:`!siginfo_t` structure with the following attributes:
   
   * :attr:`!si_pid` (process ID)
   * :attr:`!si_uid` (real user ID of the child)
   * :attr:`!si_signo` (always :data:`~signal.SIGCHLD`)
   * :attr:`!si_status` (the exit status or signal number, depending on :attr:`!si_code`)
   * :attr:`si_code` (see :data:`CLD_EXITED` for possible values)
```

* Structure this as a bulleted list, which is a lot easier to read and quickly consult, as appropriate for a reference doc
* Make it more clear/less syntactically ambiguous that "with" refers to the object, not the structure or the data contained in it.
* De-resolve these attribute reference and the `:c:type:` reference, since they aren't actually documented as such so resolution fails, and results in warnings

   If :data:`WNOHANG` is specified and there are no matching children in the

>    requested state, ``None`` is returned.  If there are no matching children

Reviewer 2: ```suggestion
   requested state, ``None`` is returned. 
   Otherwise, if there are no matching children
```

Add "otherwise", to clarify the relationship between the two (assuming that is correct), which took some time to parse as a reader.

   that could be waited for, :exc:`ChildProcessError` is raised.

   .. availability:: Unix, not Emscripten, not WASI.

   .. versionadded:: 3.3

> .. function:: waitpid(pid, options, /)

Reviewer 2: I can't reach it with a comment/suggestion, but while the prose documentation here and above states that `waitpid` is supported on both *nix and Windows (just with somewhat different behavior), the `availability` states that it is only on Unix. It seems the latter should be updated to `Unix, Windows, not Emscripten, not WASI.`

Reviewer 2: Also, the semantics of *options* are not described here, despite `wait3` and `wait4` explicitly referring to this function for a description of that argument. Presumably, it should be similar in form to the one in the previous `waitid` function, listing the flags that are required and optional (`WCONTINUED`, `WUNTRACED`, and `WNOHANG` are the available options; I'm guessing maybe at least one of the first tow is required, and the last one is optional like it is with `waitid`?)

Author: Yeah, will update the availability. `options` is described at least in that it should be `0` normally, which means no flags are required.

   number is zero); the high bit of the low byte is set if a core file was
   produced.

   If there are no children that could be waited for, :exc:`ChildProcessError`
   is raised.

   :func:`waitstatus_to_exitcode` can be used to convert the exit status into an
   exit code.

   .. availability:: Unix, not Emscripten, not WASI.

   .. seealso::

      The other :func:`!wait*` functions documented below can be used to wait for the
      completion of a specific child process and have more options.
      :func:`waitpid` is the only one also available on Windows.

.. function:: waitid(idtype, id, options, /)

   Wait for the completion of a child process.

>    *idtype* can be :data:`P_PID`, :data:`P_PGID`, :data:`P_ALL`, or, on Linux, :data:`P_PIDFD`.

Reviewer 2: ```suggestion
   *idtype* can be :data:`P_PID`, :data:`P_PGID`, :data:`P_ALL`, or (on Linux) :data:`P_PIDFD`.
```

Minor tweak to make the syntax a little more clear and avoid overloading the comma here.

   The interpretation of *id* depends on it; see their individual descriptions.

   *options* is an OR combination of flags.  At least one of :data:`WEXITED`,
   :data:`WSTOPPED` or :data:`WCONTINUED` is required;
   :data:`WNOHANG` and :data:`WNOWAIT` are additional optional flags.

   The return value is an object representing the data contained in the
   :c:type:`!siginfo_t` structure with the following attributes:

   * :attr:`!si_pid` (process ID)
   * :attr:`!si_uid` (real user ID of the child)
   * :attr:`!si_signo` (always :data:`~signal.SIGCHLD`)
   * :attr:`!si_status` (the exit status or signal number, depending on :attr:`!si_code`)

>    * :attr:`si_code` (see :data:`CLD_EXITED` for possible values)

Reviewer 2: ```suggestion
   * :attr:`!si_code` (see :data:`CLD_EXITED` for possible values)
```

Looks like this got missed somehow

   ``-1``, status is requested for any process in the process group ``-pid`` (the
   absolute value of *pid*).

   If :data:`WNOHANG` is specified and there are no matching children in the

>    requested state, ``(0, 0)`` is returned.  If there are no matching children

Reviewer 2: ```suggestion
   requested state, ``(0, 0)`` is returned.
   Otherwise, if there are no matching children
```

Likewise, make the relationship between these two a little more explicit and easier for the reader to parse.

.. function:: wait4(pid, options)

   Similar to :func:`waitpid`, except a 3-element tuple, containing the child's
   process id, exit status indication, and resource usage information is
   returned.  Refer to :func:`resource.getrusage` for details on resource usage
   information.  The arguments to :func:`wait4` are the same as those provided
   to :func:`waitpid`.

   :func:`waitstatus_to_exitcode` can be used to convert the exit status into an
   exitcode.

   .. availability:: Unix, not Emscripten, not WASI.

.. data:: P_PID
          P_PGID
          P_ALL
          P_PIDFD

   These are the possible values for *idtype* in :func:`waitid`. They affect
   how *id* is interpreted:

   * :data:`P_PID` - wait for the child whose PID is *id*.
   * :data:`P_PGID` - wait for any child whose progress group ID is *id*.
   * :data:`P_ALL` - wait for any child; *id* is ignored.

>    * :data:`P_PIDFD` - wait for the child identified by the file descriptor

Reviewer 2: ```suggestion
   * :data:`!P_PID` - wait for the child whose PID is *id*.
   * :data:`!P_PGID` - wait for any child whose progress group ID is *id*.
   * :data:`!P_ALL` - wait for any child; *id* is ignored.
   * :data:`!P_PIDFD` - wait for the child identified by the file descriptor
```

These just refer to this very same entry, so are redundant and can just have resolution disabled, per current agreed convention.

     *id* (a process file descriptor created with :func:`pidfd_open`).

   .. availability:: Unix, not Emscripten, not WASI.

>    .. note:: ``P_PIDFD` is only available on Linux >= 5.4.

Reviewer 2: ```suggestion
   .. note:: :data:`!P_PIDFD` is only available on Linux >= 5.4.
```

Fix the build error and add the appropriate syntax; I have a review in-work for the rest of it.

>    .. note:: ``P_PIDFD`` is only available on Linux >= 5.4.

Reviewer 2: ```suggestion
   .. note:: :data:`!P_PIDFD` is only available on Linux >= 5.4.
```

Be consistent with formatting/semantics as the usages above, and with general modern convention.

   .. versionadded:: 3.3
   .. versionadded:: 3.9

>       The ``P_PIDFD`` constant.

Reviewer 2: ```suggestion
      The :data:`!P_PIDFD` constant.
```

Likewise

.. data:: WCONTINUED

>    This option for :func:`waitpid`, :func:`wait3`, :func:`wait4`, and

Reviewer 2: ```suggestion
   This *options* flag for :func:`waitpid`, :func:`wait3`, :func:`wait4`, and
```

Make it more explicit what argument this is passed to, and use precise terminology to make explicit this is a flag value.

   :func:`waitid` causes child processes to also be reported if they have been

>    continued from a job control stop since their status was last reported.

Reviewer 2: ```suggestion
   :func:`waitid` causes child processes to be reported if they have been
   continued from a job control stop since they were last reported.
```

I found the syntax here quite hard to parse and understand, and simplified it a bit. Maybe it can be simplified further?

   .. availability:: Unix, not Emscripten, not WASI.

.. data:: WEXITED

>    This option for :func:`waitid` causes child processes that have terminated to

Reviewer 2: ```suggestion
   This *options* flag for :func:`waitid` causes child processes that have terminated to
```

Clarify as above

   be reported.

>    The other ``wait*`` function always report children that have terminated,

Reviewer 2: ```suggestion
   The other ``wait*`` functions always report children that have terminated,
```

Fix grammar. You might also want to consider `` :func:`!wait*` `` or `` :samp:`wait{*}` `` for the syntax, to be more consistent with the formatting/semantics of individual function names and/or emphasize the variable portion more clearly.

   so this option is not available for them.

   .. availability:: Unix, not Emscripten, not WASI.

   .. versionadded:: 3.3

.. data:: WSTOPPED

>    This option for :func:`waitid` causes child processes that have been stopped

Reviewer 2: ```suggestion
   This *options* flag for :func:`waitid` causes child processes that have been stopped
```

Clarify as above

   by the delivery of a signal to be reported.

   This option is not available for the other ``wait*`` functions.

   .. availability:: Unix, not Emscripten, not WASI.

   .. versionadded:: 3.3

.. data:: WUNTRACED

>    This option for :func:`waitpid`, :func:`wait3`, and :func:`wait4` causes

Reviewer 2: ```suggestion
   This *options* flag for :func:`waitpid`, :func:`wait3`, and :func:`wait4` causes
```

Clarity as above

   child processes to also be reported if they have been stopped but their
   current state has not been reported since they were stopped.

   This option is not available for :func:`waitid`.

   .. availability:: Unix, not Emscripten, not WASI.

.. data:: WNOHANG

>    The option for :func:`waitpid`, :func:`wait3`, :func:`wait4`, and

Reviewer 2: ```suggestion
   This *options* flag causes :func:`waitpid`, :func:`wait3`, :func:`wait4`, and
```

Clarify as above, and fix grammar

>    :func:`waitid` to return immediately if no child process status is available
>    immediately.

Reviewer 2: To avoid the jarring repetition of "immediately", a few options:

```suggestion
   :func:`waitid` to return without waiting if no child process status is available
   immediately.
```

```suggestion
   :func:`waitid` to return right away if no child process status is available
   immediately.
```

```suggestion
   :func:`waitid` to return immediately away if no child process status is available
   right away.
```

```suggestion
   :func:`waitid` to return immediately if no child process status is available
   at the time the functions are called.
```

Or, if none of these work, you could at least do

```suggestion
   :func:`waitid` to return immediately if no child process status is
   immediately available.
```

   .. availability:: Unix, not Emscripten, not WASI.

.. data:: WNOWAIT

>    The option for :func:`waitid` to leave the child in a waitable state, so that

Reviewer 2: ```suggestion
   This *options* flag causes :func:`waitid` to leave the child in a waitable state, so that
```

As above, fix grammar and clarify

>    a later wait call can be used again to retrieve the child status information.

Reviewer 2: Fix grammar, and I was unclear on the intended meaning here with more explicit syntax—do you mean a later `os.waitid` call?

```suggestion
   a later :func:`!waitid` call can be used to retrieve the child status information again.
```

...a later call to any `wait*` function?

```suggestion
   a later :func:`!wait*` call can be used to retrieve the child status information again.
```

...or something else, e.g. a later call to the underlying OS API (which could be stated more clearly)?

   This option is not available for the other ``wait*`` functions.

   .. availability:: Unix, not Emscripten, not WASI.

.. data:: CLD_EXITED
          CLD_KILLED
          CLD_DUMPED
          CLD_TRAPPED
          CLD_STOPPED
          CLD_CONTINUED

>    These are the possible values for :attr:`si_code` in the result returned by

Reviewer 2: ```suggestion
   These are the possible values for :attr:`!si_code` in the result returned by
```

Like above, this isn't an actual documented attribute (and not module-level data), so we shouldn't try to resolve it, fail and issue a warning.
